x-pack/plugins/fleet/server/services/secrets.ts

```diff
@@ -157,7 +249,49 @@ export async function extractAndWriteSecrets(opts: {
 
   return {
     packagePolicy: policyWithSecretRefs,
-    secret_references: secrets.map(({ id }) => ({ id })),
+    secretReferences: secrets.map(({ id }) => ({ id })),
+  };
+}
+
+export async function extractAndUpdateSecrets(opts: {
+  oldPackagePolicy: PackagePolicy;
+  packagePolicyUpdate: UpdatePackagePolicy;
+  packageInfo: PackageInfo;
+  esClient: ElasticsearchClient;
+}): Promise<{
+  packagePolicyUpdate: UpdatePackagePolicy;
+  secretReferences: PolicySecretReference[];
+  secretsToDelete: PolicySecretReference[];
+}> {
+  const { oldPackagePolicy, packagePolicyUpdate, packageInfo, esClient } = opts;
```

Comment:
This function is the core of the update flow. We extract the secrets from theo ld policy and the new policy, giving us two arrays of secret values and their paths.

We then get the diff of these arrays, we need to know if there are updated secrets which need to be created (`toCreate`), secrets which are no longer used (`toDelete`) and unchanged secrets (`noChange`)

For example, I have a policy with a mongo password, the password is a secret reference with ID 1. We update the password to newValue. 

We need `toCreate` a secret for newValue, `toDelete` the old secret with ID 1 (if it isn't used by any other policies), and `noChange` would be empty.